peer address update checks

thomas-fossati opened the pull request:
A first attempt to address #64 , as per discussion with Achim.

thomas-fossati committed:
peer address update checks

ekr commented:
I think the first paragraph of this PR is fine, but the rest is out of scope.

The agreed upon intent of this draft is simply to allow flow identification, not to provide a generic address migration facility. That is a complicated topic that should be out of scope here.

boaks commented on draft-ietf-tls-dtls-connection-id.md:
```diff
@@ -345,6 +345,30 @@ data the following modification is made to the additional data calculation.
+Immediately after reception, the recipient SHOULD initiate address validation
+using a Heartbeat {!RFC6520} based ping-pong with the new peer address.  This
+is to ensure that address validation can be cryptographically verified.  Both
+Heartbeat messages (request and response) MUST be wrapped into tls12_cid
```
FMPOV, **MUST** for `request` is (currently) not possible, because the use of tls12_cid is bound to a none empty cid. Only the response can considered to be MUST tls12_cid, because the cause of this address-issues is bound to use such a none empty-cid for incoming messages.

boaks commented on draft-ietf-tls-dtls-connection-id.md:
```diff
@@ -345,6 +345,30 @@ data the following modification is made to the additional data calculation.
+is to ensure that address validation can be cryptographically verified.  Both
+Heartbeat messages (request and response) MUST be wrapped into tls12_cid
+records.  If CID has not been negotiated, the sender MUST use a 0-length CID
+instead of the RFC 6347-defined record format. While the address validation
```
My feeling is, tls12_cid with empty cid was discussed some months ago. To change this, would cause changes in other parts (MAC), and I'm not sure, if this pays off. Assuming, that the response will be a tls12_cid, which risk do you see, which cause the MUST for the request?

boaks commented:
I would prefer / appreciate, if the parts after the first paragraph would start with explaining the vulnerability and the attack, which may use it. That would setup the right scope for understanding, when a mitigation is required to protect other systems.

In the discussion #64 I tried to point out, that I would prefer to leave the mitigation to the using system and just mention the "heartbeat" as one solution.
As I wrote in a comment above, I'm not sure, if introducing "tls12_cid for empty cid" will pay off. May be you can provide us the scenario, where the request requires tls12_cid as well.

thomas-fossati committed:
Addressing EKR’s comment.

- Remove heartbeat based address validation -- I’ll probably write
  up the the proposal in a separate draft;

- Push peer validation up in the stack by RECOMMENDing that
  implementations expose peer address updates to their user.

jsalowey commented:
We don't have consensus for all of this text.  I think the proposal is to address tis in a separate document.  Can we close this PR?

thomas-fossati commented:
> We don't have consensus for all of this text. I think the proposal is to address tis in a separate document. Can we close this PR?

The current version of this PR has expunged the text that outlined an address validation sub-protocol.  This can and will be sketched separately.  What is left here is just a description of what the problem is and what a receiver must do to counter at least two possible attacks using simple unilateral measures.  I'm not sure what is controversial?

boaks commented:
I would prefer, if the risk of the "man-in-the-middle" explicitly states, that

- it's a risk to be misused to attack other systems
  (the own system may attacked by such a  "man-in-the-middle" much harder with DoS by just dropping the messages)
- therefore only applies to large response messages

In my opinion, too general descriptions of risks tend to be misunderstood.

hannestschofenig commented:
This PR has been replaced by text offered in https://github.com/tlswg/dtls-conn-id/pull/68
